File: smash-ranks-android/app/src/main/java/com/garpr/android/sync/RankingsPollingSyncManagerImpl.java
package com.garpr.android.sync;

import android.support.annotation.NonNull;

import com.garpr.android.misc.GoogleApiWrapper;
import com.garpr.android.misc.Timber;
import com.garpr.android.models.PollFrequency;
import com.garpr.android.preferences.RankingsPollingPreferenceStore;
import com.google.android.gms.gcm.GcmNetworkManager;
import com.google.android.gms.gcm.PeriodicTask;
import com.google.android.gms.gcm.Task;

import java.util.concurrent.TimeUnit;

public class RankingsPollingSyncManagerImpl implements RankingsPollingSyncManager {

    private static final String TAG = "RankingsPollingSyncManagerImpl";

    private final GoogleApiWrapper mGoogleApiWrapper;
    private final RankingsPollingPreferenceStore mRankingsPollingPreferenceStore;
    private final Timber mTimber;


    public RankingsPollingSyncManagerImpl(@NonNull final GoogleApiWrapper googleApiWrapper,
            @NonNull final RankingsPollingPreferenceStore rankingsPollingPreferenceStore,
            @NonNull final Timber timber) {
        mGoogleApiWrapper = googleApiWrapper;
        mRankingsPollingPreferenceStore = rankingsPollingPreferenceStore;
        mTimber = timber;
    }

    private void disable() {
        final GcmNetworkManager gcmNetworkManager = mGoogleApiWrapper.getGcmNetworkManager();
        gcmNetworkManager.cancelAllTasks(RankingsPollingService.class);
        mTimber.d(TAG, "Sync has been disabled");
    }

    private void enable() {
        if (!mGoogleApiWrapper.isGooglePlayServicesAvailable()) {
            mTimber.w(TAG, "failed to schedule sync because Google Play Services are unavailable");
            return;
        }

        final PeriodicTask.Builder builder = new PeriodicTask.Builder()
                .setPersisted(true)
                .setService(RankingsPollingService.class)
                .setTag(TAG)
                .setUpdateCurrent(true);

        final PollFrequency pollFrequency = mRankingsPollingPreferenceStore.getPollFrequency().get();
        // noinspection ConstantConditions
        builder.setPeriod(TimeUnit.MILLISECONDS.toSeconds(pollFrequency.getTimeInMillis()));

        if (Boolean.TRUE.equals(mRankingsPollingPreferenceStore.getChargingRequired().get())) {
            builder.setRequiresCharging(true);
        } else {
            builder.setRequiresCharging(false);
        }

        if (Boolean.TRUE.equals(mRankingsPollingPreferenceStore.getWifiRequired().get())) {
            builder.setRequiredNetwork(Task.NETWORK_STATE_UNMETERED);
        } else {
            builder.setRequiredNetwork(Task.NETWORK_STATE_CONNECTED);
        }

        final PeriodicTask periodicTask = builder.build();
        final GcmNetworkManager gcmNetworkManager = mGoogleApiWrapper.getGcmNetworkManager();
        gcmNetworkManager.schedule(periodicTask);

        mTimber.d(TAG, "Sync has been enabled");
    }

    @Override
    public void enableOrDisable() {
        if (Boolean.TRUE.equals(mRankingsPollingPreferenceStore.getEnabled().get())) {
            enable();
        } else {
            disable();
        }
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/sync/RankingsPollingSyncManagerImpl.kt
package com.garpr.android.sync

import com.firebase.jobdispatcher.Constraint
import com.firebase.jobdispatcher.Lifetime
import com.firebase.jobdispatcher.RetryStrategy
import com.firebase.jobdispatcher.Trigger
import com.garpr.android.misc.FirebaseApiWrapper
import com.garpr.android.misc.GoogleApiWrapper
import com.garpr.android.misc.Timber
import com.garpr.android.models.PollFrequency
import com.garpr.android.preferences.RankingsPollingPreferenceStore

class RankingsPollingSyncManagerImpl(
        private val firebaseApiWrapper: FirebaseApiWrapper,
        private val googleApiWrapper: GoogleApiWrapper,
        private val rankingsPollingPreferenceStore: RankingsPollingPreferenceStore,
        private val timber: Timber
) : RankingsPollingSyncManager {

    companion object {
        private const val TAG = "RankingsPollingSyncManagerImpl"
    }

    private fun disable() {
        firebaseApiWrapper.getJobDispatcher().cancel(TAG)
        timber.d(TAG, "sync has been disabled")
    }

    private fun enable() {
        if (!googleApiWrapper.isGooglePlayServicesAvailable) {
            timber.w(TAG, "failed to schedule sync because Google Play Services are unavailable")
            return
        }

        val jobDispatcher = firebaseApiWrapper.getJobDispatcher()

        val jobBuilder = jobDispatcher.newJobBuilder()
                .setLifetime(Lifetime.FOREVER)
                .setRecurring(true)
                .setReplaceCurrent(true)
                .setRetryStrategy(RetryStrategy.DEFAULT_EXPONENTIAL)
                .setService(RankingsPollingJobService::class.java)
                .setTag(TAG)

        if (isChargingRequired) {
            jobBuilder.addConstraint(Constraint.DEVICE_CHARGING)
        }

        if (isWifiRequired) {
            jobBuilder.addConstraint(Constraint.ON_UNMETERED_NETWORK)
        } else {
            jobBuilder.addConstraint(Constraint.ON_ANY_NETWORK)
        }

        val pollFrequency = rankingsPollingPreferenceStore.pollFrequency.get() ?: PollFrequency.DAILY
        jobBuilder.trigger = Trigger.executionWindow(0, pollFrequency.timeInSeconds.toInt())
        jobDispatcher.mustSchedule(jobBuilder.build())

        timber.d(TAG, "sync has been enabled")
    }

    override fun enableOrDisable() {
        if (isEnabled) {
            enable()
        } else {
            disable()
        }
    }

    override var isChargingRequired: Boolean
        get() = rankingsPollingPreferenceStore.chargingRequired.get() == true
        set(value) {
            rankingsPollingPreferenceStore.chargingRequired.set(value)
            enableOrDisable()
        }

    override var isEnabled: Boolean
        get() = rankingsPollingPreferenceStore.enabled.get() == true
        set(value) {
            rankingsPollingPreferenceStore.enabled.set(value)
            enableOrDisable()
        }

    override var isWifiRequired: Boolean
        get() = rankingsPollingPreferenceStore.wifiRequired.get() == true
        set(value) {
            rankingsPollingPreferenceStore.wifiRequired.set(value)
            enableOrDisable()
        }

}
